### PR TITLE
Steel: Changes in support of a PCM-based memory model

### PR DESCRIPTION
**Pull request by nikswamy**
This is a large diff.

Most of the changes are in ulib/experimental/Steel.*. 

But, there is also a dependence on PR #2047 , which I will try to merge separately before this PR.

Also, there are some changes to modules in ulib/FStar ... 
- revising and extending FStar.IndefiniteDescription, so that it's defined in Tot for erased types and props
- notably a new FStar.PCM module, with generic definitions of PCMs.
- And some utiliies in FStar.Universe

I don't necessarily expect a detailed code review, though of course I'd be happy for any comments. But, more generally, I'd be curious to hear if there's any reason to not merge steel to master.

### CODE REVIEW

File: .docker/build/config.json [hunk 1]
```diff
@@ -20,8 +20,8 @@
     "HasLogsToExtract" : true,
 
     "NotificationEnabled" : true,
-    "NotificationChannel" : "#fstar-build",
-    "PublicBranches" : [ "master" ],
+    "NotificationChannel" : "#steel-build",
+    "PublicBranches" : [ "steel" ],
```
Comment (mtzguido) [inline .docker/build/config.json:24]: This shouldn't go to master.

File: ulib/experimental/Steel.Effect.fst [hunk 2]
```diff
@@ -18,56 +18,36 @@ module Steel.Effect
...
- *
- * Also see Steel.Effect.M.fst for an alternate attempt that chooses the
- * action trees as the underlying representation
- *)
+let join_preserves_interp hp m0 m1
+  = intro_emp m1;
+    intro_star hp emp m0 m1;
+    affine_star hp emp (join m0 m1)
+
+#push-options "--query_stats"
```
Comment (denismerigoux) [inline ulib/experimental/Steel.Effect.fst:32]: There's a `query_stats` left here

File: ulib/experimental/Steel.Effect.fst [hunk 3]
```diff
@@ -79,407 +59,127 @@ let interp_depends_only_on_post (#a:Type) (hp:a -> hprop)
...
+  = FStar.Monotonic.Pure.wp_monotonic_pure ();
+    fun m0 ->
+      let x = f () in
+      g x m0
 #pop-options
 
-polymonadic_bind (Steel, PURE) |> Steel = bind_steel_pure
+#push-options "--z3rlimit 200 --fuel 0 --ifuel 0"
+let bind_steel_pure a b pre_f post_f req_f ens_f wp_g f g
+  = admit(); //We don't really need this direction anymore and it takes ages to prove (not sure why)
```
Comment (denismerigoux) [inline ulib/experimental/Steel.Effect.fst:96]: Can we remove it if we don't need it ?

File: ulib/experimental/Steel.Effect.fsti [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+open Steel.Memory
+
+val join_preserves_interp (hp:slprop) (m0 m1:mem)
+  : Lemma
+    (requires (interp hp m0 /\ disjoint m0 m1))
+    (ensures (interp hp (join m0 m1)))
+    [SMTPat (interp hp (join m0 m1))]
+
+val respects_fp (#fp:slprop) (p: hmem fp -> prop) : prop
+#push-options "--query_stats"
```
Comment (denismerigoux) [inline ulib/experimental/Steel.Effect.fsti:31]: `query_stats` left here

File: ulib/experimental/Steel.Heap.fst [hunk 5]
```diff
@@ -0,0 +1,1080 @@
...
+   limitations under the License.
+*)
+module Steel.Heap
+module F = FStar.FunctionalExtensionality
+open FStar.FunctionalExtensionality
+open FStar.PCM
+
+#set-options "--fuel 1 --ifuel 1"
+
+// In the future, we may have other cases of cells
```
Comment (denismerigoux) [inline ulib/experimental/Steel.Heap.fst:23]: No we won't, at least not here, we can remove the commment

File: ulib/experimental/Steel.HigherReference.fst [hunk 6]
```diff
@@ -15,423 +15,281 @@
...
+open Steel.Memory
+open FStar.Ghost
+open FStar.Real
+open FStar.PCM
+open Steel.FractionalPermission
+module Atomic = Steel.Effect.Atomic
+module SB = Steel.SteelT.Basics
+
+let fractional (a:Type u#1) = option (a & perm)
+#push-options "--query_stats"
```
Comment (denismerigoux) [inline ulib/experimental/Steel.HigherReference.fst:29]: `query_stats` left here

File: ulib/experimental/Steel.HigherReference.fst [hunk 7]
```diff
@@ -15,423 +15,281 @@
...
-       interp (pts_to_ref r q v `star` pure (witnessed r fact)) m
-     ) =
-       lem_star_pure (pts_to_ref r q v) (witnessed r fact)
+  : SteelT unit (pts_to r full_perm v) (fun _ -> emp)
+  = let v_old : erased (fractional a) = Ghost.hide (Some (Ghost.reveal v, full_perm)) in
+    elim_perm_ok_star full_perm;
+    Steel.Effect.free r v_old;
+    drop _
+
+(* move these to Mem *)
```
Comment (denismerigoux) [inline ulib/experimental/Steel.HigherReference.fst:202]: Do you want ot move these now ?

File: ulib/experimental/Steel.LockCoupling.fsti [hunk 8]
```diff
@@ -0,0 +1,25 @@
+module Steel.LockCoupling
+open Steel.Memory
```
Comment (denismerigoux) [inline ulib/experimental/Steel.LockCoupling.fsti:2]: Should it be moved to `examples/steel` ?

File: ulib/experimental/Steel.MonotonicCounter.fst [hunk 9]
```diff
@@ -0,0 +1,53 @@
+(*
```
Comment (denismerigoux) [inline ulib/experimental/Steel.MonotonicCounter.fst:1]: Move to `examples/steel` ?

File: ulib/experimental/Steel.MonotonicHigherReference.fst [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+let pcm_history_preorder #a #p : Preorder.preorder (history a p) =
+  fun h0 h1 ->
+    match h0, h1 with
+    | Witnessed vh0, Witnessed vh1
+    | Current vh0 _, Witnessed vh1
+    | Witnessed vh0, Current vh1 _
+    | Current vh0 _, Current vh1 _ ->
+      vh1 `Q.extends` vh0
+
+#push-options "--max_fuel 1 --initial_fuel 1 --max_ifuel 1 --z3rlimit_factor 4 --query_stats --z3cliopt smt.qi.eager_threshold=100"
```
Comment (denismerigoux) [inline ulib/experimental/Steel.MonotonicHigherReference.fst:110]: `query_stats` left here

File: ulib/experimental/Steel.Primitive.ForkJoin.fst [hunk 11]
```diff
@@ -18,71 +18,73 @@ module Steel.Primitive.ForkJoin
...
-  : hprop
+let lock_inv_pred (r:ref bool) (p:slprop) (v:bool) : slprop =
+  pts_to r full_perm v `star` maybe_p p v
+
+let lock_inv (r:ref bool) (p:slprop)
+  : slprop
   = h_exists (lock_inv_pred r p)
 
-#set-options "--print_universes --print_implicits"
+//#set-options "--print_universes --print_implicits"
```
Comment (denismerigoux) [inline ulib/experimental/Steel.Primitive.ForkJoin.fst:41]: Line can be removed

File: ulib/experimental/Steel.ReshuffleFailure.fst [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+module SB = Steel.SteelT.Basics
+module ST = Steel.Memory.Tactics
+
+assume
+val pp : int -> slprop u#1
+
+assume
+val act (a b:slprop)
+  : SteelT int (a `star` b) (fun x -> a `star` pp x)
+
```
Comment (denismerigoux) [inline ulib/experimental/Steel.ReshuffleFailure.fst:14]: Move that to `examples/steel` ?

**Pull request merged**